File: src/main/scala/uk/yermak/audiobookconverter/FFMpegNativeConverter.scala
package uk.yermak.audiobookconverter

import java.io.{File, IOException}
import java.net.URISyntaxException
import java.util.concurrent.{Callable, CancellationException, TimeUnit}

import net.bramp.ffmpeg.progress.{Progress, TcpProgressParser}

/**
  * Created by Yermak on 29-Dec-17.
  */
object FFMpegNativeConverter {
  private val FFMPEG = new File("external/x64/ffmpeg.exe").getAbsolutePath
}

class FFMpegNativeConverter(var conversion: Conversion, var outputParameters: OutputParameters, var mediaInfo: MediaInfo, val outputFileName: String, var callback: ProgressCallback) extends Callable[ConverterOutput] {
  private var process: Process = _
  private var progressParser: TcpProgressParser = _

  @throws[IOException]
  @throws[InterruptedException]
  def convertMp3toM4a: ConverterOutput = try {
    if (conversion.getStatus.isOver) return null
    while ( {
      ProgressStatus.PAUSED == conversion.getStatus
    }) Thread.sleep(1000)
    progressParser = new TcpProgressParser((progress: Progress) => {
      def foo(progress: Progress) = {
        callback.converted(progress.out_time_ns / 1000000, progress.total_size)
        if (progress.isEnd) callback.completedConversion()
      }

      foo(progress)
    })
    progressParser.start()
    var ffmpegProcessBuilder: ProcessBuilder = null
    if (outputParameters.isAuto) if (mediaInfo.getCodec == "aac") ffmpegProcessBuilder = new ProcessBuilder(FFMpegNativeConverter.FFMPEG, "-i", mediaInfo.getFileName, "-vn", "-codec:a", "copy", "-f", "ipod", "-progress", progressParser.getUri.toString, outputFileName)
    else ffmpegProcessBuilder = new ProcessBuilder(FFMpegNativeConverter.FFMPEG, "-i", mediaInfo.getFileName, "-vn", "-codec:a", "aac", "-f", "ipod", "-progress", progressParser.getUri.toString, outputFileName)
    else ffmpegProcessBuilder = new ProcessBuilder(FFMpegNativeConverter.FFMPEG, "-i", mediaInfo.getFileName, "-vn", "-codec:a", "aac", "-f", "ipod", outputParameters.getFFMpegQualityParameter, outputParameters.getFFMpegQualityValue, "-ar", String.valueOf(outputParameters.getFFMpegFrequencyValue), "-ac", String.valueOf(outputParameters.getFFMpegChannelsValue), "-cutoff", outputParameters.getCutoffValue, "-progress", progressParser.getUri.toString, outputFileName)
    process = ffmpegProcessBuilder.start
    val ffmpegIn = process.getInputStream
    val ffmpegErr = process.getErrorStream
    StreamCopier.copy(ffmpegIn, System.out)
    StreamCopier.copy(ffmpegErr, System.err)
    var finished = false
    while ( {
      !conversion.getStatus.isOver && !finished
    }) finished = process.waitFor(500, TimeUnit.MILLISECONDS)
    new ConverterOutput(mediaInfo, outputFileName)
  } catch {
    case ce: CancellationException =>
      null
    case e: URISyntaxException =>
      throw new RuntimeException(e)
  } finally {
    Utils.closeSilently(process)
    Utils.closeSilently(progressParser)
  }

  @throws[Exception]
  override def call = convertMp3toM4a
}

File: src/main/java/uk/yermak/audiobookconverter/FFMpegNativeConverter.java
package uk.yermak.audiobookconverter;

import net.bramp.ffmpeg.progress.ProgressParser;
import net.bramp.ffmpeg.progress.TcpProgressParser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.lang.invoke.MethodHandles;
import java.net.URISyntaxException;
import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.TimeUnit;

/**
 * Created by Yermak on 29-Dec-17.
 */
public class FFMpegNativeConverter implements Callable<String> {
    final static Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
    private ConversionGroup conversionGroup;
    private MediaInfo mediaInfo;
    private final String outputFileName;
    private ProgressCallback callback;
    private Process process;
    private ProgressParser progressParser = null;


    public FFMpegNativeConverter(ConversionGroup conversionGroup, MediaInfo mediaInfo, String outputFileName, ProgressCallback callback) {
        this.conversionGroup = conversionGroup;
        this.mediaInfo = mediaInfo;
        this.outputFileName = outputFileName;
        this.callback = callback;
    }

    @Override
    public String call() throws Exception {
        try {
            if (conversionGroup.getStatus().isOver()) return null;
            while (ProgressStatus.PAUSED.equals(conversionGroup.getStatus())) Thread.sleep(1000);

            progressParser = new TcpProgressParser(progress -> {
                callback.converted(progress.out_time_ns / 1000000, progress.total_size);
                logger.debug("Completed conversion: {} from {}", progress.out_time_ns / 1000000, progress.total_size);
                if (progress.isEnd()) {
                    callback.converted(progress.out_time_ns / 1000000, progress.total_size);
                    callback.completedConversion();
                }
            });
            progressParser.start();

            ProcessBuilder ffmpegProcessBuilder;
            OutputParameters outputParameters = conversionGroup.getOutputParameters();

            if (outputParameters.needReencode(mediaInfo.getCodec())) {
                logger.debug("Re-encoding to {} for {}", outputParameters.format, outputFileName);
                ffmpegProcessBuilder = new ProcessBuilder(outputParameters.getReencodingOptions(mediaInfo, progressParser.getUri().toString(), outputFileName));
            } else {
                logger.debug("Transcoding {} stream for {}", outputParameters.format, outputFileName);
                ffmpegProcessBuilder = new ProcessBuilder(outputParameters.getTranscodingOptions(mediaInfo, progressParser.getUri().toString(), outputFileName));
            }
            process = ffmpegProcessBuilder.start();

            ByteArrayOutputStream out = new ByteArrayOutputStream();
            StreamCopier.copy(process.getInputStream(), out);
            ByteArrayOutputStream err = new ByteArrayOutputStream();
            StreamCopier.copy(process.getErrorStream(), err);

            boolean finished = false;
            while (!conversionGroup.getStatus().isOver() && !finished) {
                finished = process.waitFor(500, TimeUnit.MILLISECONDS);
            }
            logger.debug("Converter Out: {}", out.toString());
            logger.error("Converter Error: {}", err.toString());

            Mp4v2InfoLoader.updateDuration(mediaInfo, outputFileName);
            return outputFileName;
        } catch (CancellationException ce) {
            return null;
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        } finally {
            Utils.closeSilently(process);
            Utils.closeSilently(progressParser);
        }
    }


}
